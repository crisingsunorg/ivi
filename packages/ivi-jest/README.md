`ivi-jest` provides a collection of useful tools for testing with [Jest](https://jestjs.io/).

## Motivation

Jest resets global state for every test module, but all tests in a test module share the same global state. This package
provides a set of tools to make it easier to write isolated unit tests.

## API

### DOM

`useResetDOMGlobalEventListeners()` tracks event handlers for a window and document objects and removes them after
each test.

`useResetDOM()` enables following resets:

- `useResetDOMGlobalEventListeners()`

`useDOMElement()` creates a DOM element and optionally mounts it to the document body.

`useRequestAnimationFrame()` mocks `requestAnimationFrame()` and `cancelAnimationFrame()`.

### Modules

`useResetModules()` resets modules before each test.

`useModule()` imports module before each test.

### Mock

`useMockFn()` creates a mock function and resets it after each test.

`useSpyOn()` creates a spy object and resets it after each test.

### Environment

`useEnv()` overrides `process.env` before each tests and restores to the original value after each test.

### Promise

`usePromiseQueue()` adds support for waiting until promise tasks are executed.

### Timers

`useTimers()` mocks timers.

### Computed Value

`useComputedValue()` creates a value that will be recomputed before each test.
