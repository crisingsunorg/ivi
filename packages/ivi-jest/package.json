{
  "private": true,
  "name": "ivi-jest",
  "version": "0.0.1",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "src",
    "!dist/**/*.tsbuildinfo",
    "!src/**/__tests__",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json"
  },
  "dependencies": {
    "jsdom": "^14.0.0"
  },
  "peerDependencies": {},
  "devDependencies": {},
  "description": "Jest utils for ivi library",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ivi"
  ],
  "homepage": "https://github.com/localvoid/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/localvoid/ivi"
  },
  "bugs": {
    "url": "https://github.com/localvoid/ivi/issues"
  },
  "github": "https://github.com/localvoid/ivi"
}
