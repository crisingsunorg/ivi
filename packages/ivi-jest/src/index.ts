export { useComputedValue } from "./value";
export { useEnv } from "./env";
export { useMockFn, useSpyOn } from "./mock";
export { useResetModules, useModule, useIVI, useHTML, useSVG, useTest } from "./module";
export { useResetDOMGlobalEventListeners, useResetDOM, useDOMElement } from "./dom";
