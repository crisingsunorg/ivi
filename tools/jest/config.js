module.exports = {
  transform: {
    "\\.ts$": "ts-jest",
  },
  testRegex: "/__tests__/.*\\.spec\\.ts$",
  testPathIgnorePatterns: [
    "/node_modules/",
    "/dist/",
    "\\.snap$",
    "/coverage/",
  ],
  moduleFileExtensions: ["ts", "js", "json"],
  collectCoverageFrom: ["src/**/*.ts"],
  resetMocks: true,
  verbose: true,
  globals: {
    "DEBUG": true,
    "TARGET": "browser",
    "ts-jest": {
      "tsConfigFile": "./tsconfig.json",
    },
  },
};
