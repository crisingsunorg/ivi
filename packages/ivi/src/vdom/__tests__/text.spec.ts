import { useResetJSDOM, useResetModules, useDOMElement, useTest } from "ivi-jest";
import { useDOMOpsCounters } from "./jest";

useResetJSDOM();
useResetModules();
const c = useDOMElement();
const domOps = useDOMOpsCounters();
const t = useTest();
const r = (text: string | number | null) => t.render(text, c()).domNode!;

describe("mount text node", () => {
  test("empty string", () => {
    const n = r("");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string", () => {
    const n = r("abc");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("zero number", () => {
    const n = r(0);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("number", () => {
    const n = r(123);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });
});

describe("update text node", () => {
  test("null to empty string", () => {
    r(null);
    const n = r("");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("null to string", () => {
    r(null);
    const n = r("abc");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("null to zero number", () => {
    r(null);
    const n = r(0);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("null to number", () => {
    r(null);
    const n = r(123);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("empty string to null", () => {
    r("");
    const n = r(null);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("empty string to string", () => {
    r("");
    const n = r("abc");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("empty string to zero number", () => {
    r("");
    const n = r(0);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("empty string to number", () => {
    r("");
    const n = r(123);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string to null", () => {
    r("abc");
    const n = r(null);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string to empty string", () => {
    r("abc");
    const n = r("");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string to zero number", () => {
    r("abc");
    const n = r(0);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string to number", () => {
    r("abc");
    const n = r(123);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("zero number to null", () => {
    r(0);
    const n = r(null);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("zero number to string", () => {
    r(0);
    const n = r("abc");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("zero number to empty string", () => {
    r(0);
    const n = r("");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("zero number to number", () => {
    r(0);
    const n = r(123);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("number to null", () => {
    r(123);
    const n = r(null);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("number to string", () => {
    r(123);
    const n = r("abc");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("number to empty string", () => {
    r(123);
    const n = r("");
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("number to zero number", () => {
    r(123);
    const n = r(0);
    expect(n).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });

  test("string to fragment", () => {
    r("abc");
    const { domNode } = t.render([1, 2, 3], c());
    expect(domNode).toMatchSnapshot();
    expect(domOps()).toMatchSnapshot();
  });
});
