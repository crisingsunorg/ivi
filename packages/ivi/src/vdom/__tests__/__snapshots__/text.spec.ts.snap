// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount text node empty string 1`] = ``;

exports[`mount text node empty string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`mount text node number 1`] = `123`;

exports[`mount text node number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`mount text node string 1`] = `abc`;

exports[`mount text node string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`mount text node zero number 1`] = `0`;

exports[`mount text node zero number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node empty string to null 1`] = `null`;

exports[`update text node empty string to null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node empty string to number 1`] = `123`;

exports[`update text node empty string to number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node empty string to string 1`] = `abc`;

exports[`update text node empty string to string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node empty string to zero number 1`] = `0`;

exports[`update text node empty string to zero number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node null to empty string 1`] = ``;

exports[`update text node null to empty string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node null to number 1`] = `123`;

exports[`update text node null to number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node null to string 1`] = `abc`;

exports[`update text node null to string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node null to zero number 1`] = `0`;

exports[`update text node null to zero number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node number to empty string 1`] = ``;

exports[`update text node number to empty string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node number to null 1`] = `null`;

exports[`update text node number to null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node number to string 1`] = `abc`;

exports[`update text node number to string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node number to zero number 1`] = `0`;

exports[`update text node number to zero number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node string to empty string 1`] = ``;

exports[`update text node string to empty string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node string to fragment 1`] = `1`;

exports[`update text node string to fragment 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 4,
  "innerHTML": 0,
  "insertBefore": 4,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node string to null 1`] = `null`;

exports[`update text node string to null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node string to number 1`] = `123`;

exports[`update text node string to number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node string to zero number 1`] = `0`;

exports[`update text node string to zero number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node zero number to empty string 1`] = ``;

exports[`update text node zero number to empty string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node zero number to null 1`] = `null`;

exports[`update text node zero number to null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node zero number to number 1`] = `123`;

exports[`update text node zero number to number 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`update text node zero number to string 1`] = `abc`;

exports[`update text node zero number to string 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 1,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 1,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;
